server/commands/connected_objects.go

```diff
@@ -0,0 +1,355 @@
+// limit. If a 0 limit is provided then all objects of the provided objectType will be
+// returned.
+func (c *ConnectedObjectsCommand) StreamedConnectedObjects(
+	ctx context.Context,
+	req *ConnectedObjectsRequest,
+	resultChan chan<- string, // object string (e.g. document:1)
+) error {
+
+	var foundCount *uint32
+	if req.Limit > 0 {
+		foundCount = new(uint32)
+	}
+
+	var foundObjects sync.Map
+	return c.streamedConnectedObjects(ctx, req, resultChan, &foundObjects, foundCount)
+}
+
+type reverseExpandRequest struct {
+	storeID          string
+	ingress          *graph.RelationshipIngress
+	sourceObjectRef  *openfgapb.RelationReference
+	targetUserRef    *openfgapb.ObjectRelation
+	contextualTuples []*openfgapb.TupleKey
+	limit            uint32
+}
+
+func (c *ConnectedObjectsCommand) reverseExpandTupleToUserset(
+	ctx context.Context,
+	req *reverseExpandRequest,
+	resultChan chan<- string,
+	foundObjectsMap *sync.Map,
+	foundCount *uint32,
+) error {
+
+	store := req.storeID
+
+	ingress := req.ingress.Ingress
+
+	sourceObjectType := req.sourceObjectRef.GetType()
+	sourceObjectRel := req.sourceObjectRef.GetRelation()
+
+	tuplesetRelation := req.ingress.TuplesetRelation.GetRelation()
+
+	var tuples []*openfgapb.Tuple
+	for _, t := range req.contextualTuples {
+
+		object := t.GetObject()
+		objectType, _ := tuple.SplitObject(object)
+		if objectType != ingress.GetType() {
+			continue
+		}
+
+		if t.GetRelation() != tuplesetRelation {
+			continue
+		}
+
+		targetUserStr := req.targetUserRef.GetObject()
+
+		userObj, _ := tuple.SplitObjectRelation(t.GetUser())
+		if userObj == targetUserStr || userObj == "*" {
+			tuples = append(tuples, &openfgapb.Tuple{Key: t})
+		}
+	}
+	iter1 := storage.NewStaticTupleIterator(tuples)
+
+	iter2, err := c.Datastore.ReadStartingWithUser(ctx, store, storage.ReadStartingWithUserFilter{
+		ObjectType: req.ingress.Ingress.GetType(),
+		Relation:   tuplesetRelation,
+		UserFilter: []*openfgapb.ObjectRelation{
+			{Object: req.targetUserRef.Object},
+			{Object: "*"},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	iter := storage.NewCombinedIterator(iter1, iter2)
+
+	g := errgroup.Group{}
+	g.SetLimit(100) // set some concurrency limit
+
+	for {
+		t, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, storage.ErrIteratorDone) {
+				break
+			}
+
+			return err
+		}
+
+		tk := t.GetKey()
+
+		foundObject := tk.GetObject()
+		foundObjectType, _ := tuple.SplitObject(foundObject)
+
+		if _, ok := foundObjectsMap.Load(foundObject); ok {
+			// todo(jon-whit): we could optimize this by avoiding reading this
+			// from the database in the first place
+
+			// if we've already evaluated/found the object, then continue
+			continue
+		}
+
+		if foundObjectType == sourceObjectType {
+			if foundCount != nil && atomic.LoadUint32(foundCount) >= req.limit {
+				break
+			}
+
+			if foundCount != nil {
+				atomic.AddUint32(foundCount, 1)
+			}
```

Comment:
Isn't this prone to the same race condition that we had found in ListObjects v0?

```suggestion
if atomic.AddUint32(foundCount, 1) > req.Limit {
	break
}
```

Reply:
Ah, I updated this in one spot but not here. It's supposed to be

```
if foundCount != nil && atomic.AddUint32(foundCount, 1) > req.limit { ... }
```

will fix this occurance.

Reply:
This is resolved now.